worked out most of the issues form #9

wegenmat opened the pull request:
- moved lightcurves to corresponding paper folder (now done automatically by the script),
- changed source id to integers only,
- renamed the existing files,
- added 'none' as paper-format for non-ADS references

cdeil commented:
> added 'none' as paper-format for non-ADS references

None has a special meaning in Python and YAML as "not available".

I don't think it's a good choice here, because we will have some kind of reference record, even if it isn't an ADS bibcode.

What I'd suggest is to put
```
data_id = something
```
and then to (in a separate PR) introduce a database (YAML file) of `data_id` entries, that have a comment or URL or other info where the data comes from.

Even for cases where we don't know where the data comes from originally (if we allow that), it would be good to have a `data_id` internally and to just state that in a comment there.

cdeil commented on input/lightcurves/desy_lc.py:
```diff
@@ -105,49 +119,76 @@ def process_one_file(source_id, filename):
-        tables[x].write(source_id + '_' + filenames[x] + '.ecsv', format='ascii.ecsv')
+    # ceate folder structure and write files
+    count_files = 0
+    paper_repo = Path('/afs/ifh.de/group/amanda/scratch/wegenmat/gamma-cat/input/papers') # will be changed to 'input/papers' if the script is called from top-level repo
```
The version we commit to the repo shouldn't need hard-coded paths that e.g. I can't use.
Can you please change the script here?

cdeil commented on input/lightcurves/desy_lc.py:
```diff
@@ -105,49 +119,76 @@ def process_one_file(source_id, filename):
+    for x in range(0, len(filenames)):        
+        # create folder structure for none_paper_id
+        if tables[x].meta['data_id'] == none_paper_id:
+            if Path(str(paper_repo / none_paper_id / 'lightcurves' / 'source_id_') + source_id).exists() is False:
```
Here you're writing this three times:
```
Path(str(paper_repo / none_paper_id / 'lightcurves' / 'source_id_') + source_id)
```
To avoid code duplication, you can put a local variable and use it three times.
Also, it's possible to create the Path object in a bit simpler way:
```python
path = paper_repo / none_paper_id / 'lightcurves/source_id_{}'.format(source_id)
```

wegenmat commented:
I agree on your suggestions. I will do the minor code fixes immediately and and implement the new header keywords as discussed in issues #42 asap

cdeil commented on input/papers/1999/1999AA...350..757A/tev-000049-lc.ecsv:
```diff
@@ -0,0 +1,33 @@
+# - {name: flux, unit: Crab, datatype: float32}
+# - {name: flux_err, unit: Crab, datatype: float32}
+# meta: !!omap
+# - COMMENT: [Mrk 421]
```
My suggestion would be that we put comments like here:
https://github.com/gammapy/gamma-cat/blob/b103005c0e1f3b68c9d55f6664da6260f7ea3ecc/input/papers/2008/2008A%2526A...477..353A/tev-000093.ecsv#L12

I.e. call the field `comments`, and use one (possibly multi-line) string, not a list.

OK?

Also, if it's just about the source name, my suggestion would be that we either don't put it input files at all.
Or if you want to put it because you think it's useful for maintenance, then let's put it like this:
https://github.com/gammapy/gamma-cat/blob/b103005c0e1f3b68c9d55f6664da6260f7ea3ecc/input/papers/2008/2008A%2526A...477..353A/tev-000093.ecsv#L10

You can go to https://gammapy.github.io/gamma-cat/sources.html and find the `common_name` for your sources. In this case it's `Markarian 421`, which comes from here: https://github.com/gammapy/gamma-cat/blob/master/input/sources/tev-000049.yaml#L3

cdeil commented on input/lightcurves/desy_lc.py:
```diff
@@ -46,22 +46,34 @@ def process_one_file(source_id, filename):
+    tab['is_ul'][0:len_old] = flag
+    tab.replace_column('is_ul', tab['is_ul'][0:len_old])
+    # print(tab['is_ul'])
+
```
I think this can be simplified.

Something like this should work, no?
```python
from astropy.table import Table
table = Table()
table['fflag'] = ['<', '', '<', '', '']
table['is_ul'] = ['<' in _ for _ in table['fflag']]
del table['fflag']
```
The key is to always compute new column / row data as Python lists / dicts if strings are involved, and then let `Table` and `Column` to the work on converting to Numpy arrays under the hood.

cdeil commented:
I fixed all SEDs to encode the upper limits according to the spec earlier today.

Using the ack tool (http://beyondgrep.com/) you can get a list of example SEDs with upper limits like this:
```
$ ack dnde_ul docs/data/sources/
docs/data/sources/000024/gammacat_000024_2015Sci...347..406H_sed.ecsv
8:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
16:e_ref dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000026/gammacat_000026_2015Sci...347..406H_sed.ecsv
8:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
16:e_ref dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000027/gammacat_000027_2015Sci...347..406H_sed.ecsv
8:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
16:e_ref dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000088/gammacat_000088_2014MNRAS.439.2828A_sed.ecsv
10:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
19:e_ref e_min e_max dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000089/gammacat_000089_2014ApJ...794L...1A_sed.ecsv
10:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
17:e_ref e_min e_max dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000106/gammacat_000106_2009A%26A...503..817A_sed.ecsv
8:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
15:e_ref dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000106/gammacat_000106_2016Natur.531..476H_sed.ecsv
10:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
19:e_ref e_min e_max dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000112/gammacat_000112_2008A%26A...481..401A_sed.ecsv
8:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
16:e_ref dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000133/gammacat_000133_2016arXiv160900600H_sed.ecsv
10:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
21:e_ref e_min e_max dnde dnde_errn dnde_errp dnde_ul

docs/data/sources/000156/gammacat_000156_2008A%26A...481..401A_sed.ecsv
8:# - {name: dnde_ul, unit: 1 / (cm2 s TeV), datatype: float32}
18:e_ref dnde dnde_errn dnde_errp dnde_ul
```

Note that you need to store the upper limit values in a column `flux_ul`. That's something you have to change to be compatible with the (SED) spec, right?

For SED I decided to not put the `is_ul` column in the input files, but to use `nan` in the dnde and dnde_ul columns as appropritate.

wegenmat commented:
simplified the script and and implemented compatibility with SED specs as I understood:
set `flux_ul = flux and flux = np.nan if flux is upper limit`
else set `flux_ul = np.nan`

cdeil commented:
One more question: do you know in which "time scale" your MJD times are?

The common ones are TT and UTC, and the difference is on the order of a minute.
See https://github.com/open-gamma-ray-astro/gamma-astro-data-formats/pull/61#issuecomment-225132704

Maybe you could just add this info via the `TIMESYS` header keyword, i.e. add `TIMESYS=TT` or `TIMESYS=UTC`, whichever it is?
See https://github.com/open-gamma-ray-astro/gamma-astro-data-formats/pull/61#issuecomment-235310589

wegenmat commented:
The time scale is UTC, I added this in the `timesys` header keyword.
`.meta['COMMENT']` was deleted, since it only contains the common name of the source. I think if we want to add a common name in the lightcurve tables, this should be done automatically by a script for all lightcurves, not only the ones from DESY

cdeil commented:
@wegenmat - This is ready, right?
Should I merge it now or is there anything else you wanted to do in this branch?

wegenmat commented:
one more thing I would like to do is to update the `README`. But if you need to merge asap, don't wait for me. This can be done later on.

cdeil commented:
No need to hurry from my side. Just let me know when you'd like me to merge it.

wegenmat commented:
i am fine with merging

cdeil commented:
There's a merge conflict wrt. master now in `input/schemas/lightcurve-format.md`.

@wegenmat - Do you want to resolve it via git rebase against master? Or would you prefer I just resolve the conflict by merging this branch locally?

wegenmat commented:
I will try to rebase the branch. If that does not work, then merging locally would be appropriate

cdeil commented:
Good luck with rebasing. I've made a backup copy. :-)
I'm offline soon, will check back tonight.

wegenmat commented:
Did it work?

cdeil commented:
Everything looks good. Merging this now.

Thank you very much!!!

I'll update the Python script in `gammacat` in master today to adapt to the changes here.

We still have this issue open to continue the discussion about lightcurves in the future: https://github.com/gammapy/gamma-cat/issues/9#issuecomment-257274913
